src/pybind/mgr/dashboard/controllers/docs.py

```diff
@@ -337,7 +337,9 @@ def _gen_spec(cls, all_endpoints=False, base_url="", offline=False):
         if all_endpoints:
             base_url = ""
 
-        host = cherrypy.request.base.split('://', 1)[1] if not offline else 'example.com'
+        host = cherrypy.request.headers.get('Host')
+        if not host:
+            host = 'example.com'
```

Comment:
> Given the `X-Forwarded-*` headers are a bit sparse and non-standard, why not using the [`Forwarded`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) header (aka [PROXY protocol](http://www.haproxy.org/download/1.8/doc/proxy-protocol.txt)), that provides all the info required (origin IP, proxy hostname and scheme) in a single header.
> 
> HAProxy supports this header by enabling the PROXY protocol ([`send-proxy` flag](https://www.haproxy.com/blog/using-haproxy-with-the-proxy-protocol-to-better-secure-your-database/)).
> 
Yes, this looks great. I think it may work even with SSL passthrough too. 
> If not, you can optionally fall back to the older X-Forwarded... And why not trying this as another custom Cherrypy tool?
Actually I chose the simple config approach (set proto, host & port) bcz there maybe different ways of fetching the full url in different reverse proxies and we can't cover all of them. This approach makes it simple as the user just have to set the usual headers which can be set using any rev proxy.

Reply:
And example.com is there to pass the openapi check, as openapi.yaml uses example.com as the host